receiver/kafkareceiver/consumer_sarama.go

```diff
@@ -0,0 +1,284 @@
+// decodes telemetry data using a given unmarshaler, and passes
+// them to a consumer.
+type saramaConsumer struct {
+	config           *Config
+	topics           []string
+	settings         receiver.Settings
+	telemetryBuilder *metadata.TelemetryBuilder
+	newConsumeFn     newConsumeMessageFunc
+
+	mu                sync.Mutex
+	started           bool
+	shutdown          bool
+	consumeLoopClosed chan struct{}
+	consumerGroup     sarama.ConsumerGroup
+}
+
+func (c *saramaConsumer) Start(_ context.Context, host component.Host) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.shutdown {
+		return errors.New("kafka consumer already shut down")
+	}
+	if c.started {
+		return errors.New("kafka consumer already started")
+	}
+
+	obsrecv, err := receiverhelper.NewObsReport(receiverhelper.ObsReportSettings{
+		ReceiverID:             c.settings.ID,
+		Transport:              transport,
+		ReceiverCreateSettings: c.settings,
+	})
+	if err != nil {
+		return err
+	}
+
+	consumerGroup, err := kafka.NewSaramaConsumerGroup(
+		context.Background(),
+		c.config.ClientConfig,
+		c.config.ConsumerConfig,
+	)
+	if err != nil {
+		return err
+	}
+	c.consumerGroup = consumerGroup
+
+	handler := &consumerGroupHandler{
+		id:                c.settings.ID,
+		logger:            c.settings.Logger,
+		ready:             make(chan bool),
+		obsrecv:           obsrecv,
+		autocommitEnabled: c.config.AutoCommit.Enable,
+		messageMarking:    c.config.MessageMarking,
+		telemetryBuilder:  c.telemetryBuilder,
+		backOff:           newExponentialBackOff(c.config.ErrorBackOff),
+	}
+	consumeMessage, err := c.newConsumeFn(host, obsrecv, c.telemetryBuilder)
+	if err != nil {
+		return err
+	}
+	handler.consumeMessage = consumeMessage
+
+	c.consumeLoopClosed = make(chan struct{})
+	c.started = true
+	go c.consumeLoop(handler)
+	return nil
+}
+
+func (c *saramaConsumer) consumeLoop(handler sarama.ConsumerGroupHandler) {
+	defer close(c.consumeLoopClosed)
+
+	ctx := context.Background()
+	for {
+		// `Consume` should be called inside an infinite loop, when a
+		// server-side rebalance happens, the consumer session will need to be
+		// recreated to get the new claims
+		if err := c.consumerGroup.Consume(ctx, c.topics, handler); err != nil {
+			if errors.Is(err, sarama.ErrClosedConsumerGroup) {
+				c.settings.Logger.Info("Consumer stopped", zap.Error(ctx.Err()))
+				return
+			}
+			c.settings.Logger.Error("Error from consumer", zap.Error(err))
+		}
+	}
+}
+
+func (c *saramaConsumer) Shutdown(ctx context.Context) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.shutdown {
+		return nil
+	}
+	c.shutdown = true
+	if !c.started {
+		return nil
+	}
+
+	if err := c.consumerGroup.Close(); err != nil {
+		return err
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-c.consumeLoopClosed:
+	}
+	return nil
+}
+
+type consumerGroupHandler struct {
+	id             component.ID
+	consumeMessage consumeMessageFunc
+	ready          chan bool
+	readyCloser    sync.Once
```

Comment:
Would this be better to use https://pkg.go.dev/sync#OnceValue ?

Reply:
ready/readyCloser can actually be deleted, it's leftover code that doesn't do anything anymore